cmd/mattermost/commands/db.go

```diff
@@ -133,20 +154,103 @@ func resetCmdF(command *cobra.Command, args []string) error {
 
 func migrateCmdF(command *cobra.Command, args []string) error {
 	cfgDSN := getConfigDSN(command, config.GetEnvironment())
+	recoverFlag, _ := command.Flags().GetBool("auto-recover")
+	savePlan, _ := command.Flags().GetBool("save-plan")
+	dryRun, _ := command.Flags().GetBool("dry-run")
 	cfgStore, err := config.NewStoreFromDSN(cfgDSN, true, nil, true)
 	if err != nil {
 		return errors.Wrap(err, "failed to load configuration")
 	}
 	config := cfgStore.Get()
 
-	store := sqlstore.New(config.SqlSettings, nil)
-	defer store.Close()
+	plan, err := sqlstore.GeneratePlan(config.SqlSettings, recoverFlag)
+	if err != nil {
+		return errors.Wrap(err, "failed to generate migration plan")
+	}
+
+	if savePlan {
+		backend, err2 := filestore.NewFileBackend(config.FileSettings.ToFileBackendSettings(false, true))
+		if err2 != nil {
+			return fmt.Errorf("failed to initialize filebackend: %w", err2)
+		}
+
+		b, mErr := json.MarshalIndent(plan, "", "  ")
+		if mErr != nil {
+			return fmt.Errorf("failed to marshal plan: %w", mErr)
+		}
+
+		_, err = backend.WriteFile(bytes.NewReader(b), "migration_plan.json")
+		if err != nil {
+			return fmt.Errorf("failed to write migration plan: %w", err)
+		}
+
+		CommandPrettyPrintln("The migration plan has been saved.")
+	}
+
+	err = sqlstore.MigrateWithPlan(config.SqlSettings, plan, dryRun)
+	if err != nil {
+		return errors.Wrap(err, "failed to load configuration")
+	}
 
 	CommandPrettyPrintln("Database successfully migrated")
 
 	return nil
 }
 
+func downgradeCmdF(command *cobra.Command, args []string) error {
+	cfgDSN := getConfigDSN(command, config.GetEnvironment())
+	cfgStore, err := config.NewStoreFromDSN(cfgDSN, true, nil, true)
+	if err != nil {
+		return errors.Wrap(err, "failed to load configuration")
+	}
+	config := cfgStore.Get()
+
+	dryRun, _ := command.Flags().GetBool("dry-run")
+	recoverFlag, _ := command.Flags().GetBool("auto-recover")
+
+	backend, err2 := filestore.NewFileBackend(config.FileSettings.ToFileBackendSettings(false, true))
+	if err2 != nil {
+		return fmt.Errorf("failed to initialize filebackend: %w", err2)
+	}
+
+	// check if the input is version numbers or a file
+	// if the input is given as a file, we assume it's a migration plan
+	versions := strings.Split(args[0], ",")
```

Comment:
What does the numbers indicate here? That those numbers are the ones for which reverse migrations will run?

Reply:
nvm, saw the PR message.